route_guide/route_guide_client.cc

```diff
@@ -23,238 +23,346 @@
+                          << " seconds" << std::endl;
+              })
+              // .finished callback is invoked when the call is finished.
+              // In this callback we just start our continuation.
+              .finished([](
+                            auto&& data,
+                            auto& continuation,
+                            auto&& response) {
+                continuation.Start(grpc::Status::OK);
+              });
   }
 
-  void RouteChat() {
-    std::thread writer;
-    Notification<Status> done;
-    client_->Call<RouteGuide, Stream<RouteNote>, Stream<RouteNote>>(
-        "RouteChat",
-        [&](auto&& call, bool ok) {
-          call->OnFinished([&](auto*, const Status& status) {
-            done.Notify(status);
-          });
-          if (!ok) {
-            call->context()->TryCancel();
-            call->Finish();
-          } else {
-            call->OnRead([](auto*, auto&& server_note) {
-              if (server_note) {
+  // RouteChat() demonstrates a bidirectional streaming RPC. The client
+  // sends a sequence of 4 messages. Each message consists of a RouteNote
+  // (see protos/route_guide.proto), which contains a location. The server
+  // streams back any notes it has previously received for those locations.
+  auto RouteChat() {
+    return client_->Call<
+               RouteGuide,
+               Stream<RouteNote>,
+               Stream<RouteNote>>("RouteChat")
+        | Client::Handler()
+              .context(std::thread{})
+              // Send a stream of 4 points. We do that in a separate thread to
+              // be able to receive messages simultaneously.
+              .ready([](auto&& writer, auto& call) {
+                writer = std::thread([&]() {
+                  const std::vector<RouteNote> notes{
+                      MakeRouteNote("First message", 0, 0),
+                      MakeRouteNote("Second message", 0, 1),
+                      MakeRouteNote("Third message", 1, 0),
+                      MakeRouteNote("Fourth message", 0, 0)};
+                  for (const RouteNote& note : notes) {
+                    std::cout << "Sending message " << note.message()
+                              << " at " << note.location().latitude()
+                              << ", " << note.location().longitude()
+                              << std::endl;
+                    call.Write(note);
+                    std::this_thread::sleep_for(100ms);
+                  }
+                  call.WritesDone();
+                });
+                writer.detach();
+              })
+              // The .body() callback is invoked every time the server
+              // sends a response.
+              .body([](
+                        auto&& writer,
+                        auto& call,
+                        auto&& server_note) {
+                // We should check if the response is valid from the server.
+                if (!server_note) {
```

Comment:
As with the comments below about type-safe checks:
```suggestion
                if (server_note == nullptr) {
```